### ISSUE TITLE
Bad scoping behavior

### ISSUE THREAD

**Issue by munificent** [munificent/wren#101]
Wren has two existing problems related to scope that really need to be fixed. They are, in my mind, the biggest "broken" part of Wren and my highest priority to address.

The first is that it doesn't understand mutual recursion at the top level. This program doesn't work (where I'll define "doesn't work" in a second):

```dart
class Foo {
  bar { new Bar }
}

class Bar {
  foo { new Foo }
}
```

The second is that, within a method, any unknown identifier is presumed to be a getter on `this`. So in this program:

```dart
class Foo {
  bar { someRandomName }
}
```

It treats `someRandomName` as equivalent to `this.someRandomName`. So by doesn't work, what I mean is that the first program is actually interpreted like:

```dart
class Foo {
  bar { new this.Bar } // Bar isn't defined yet so becomes self send
}

class Bar {
  foo { new Foo }      // Foo is now, so is lexical
}
```

I think we can all agree that's pretty busted. The fact that a lexical variable shadows a getter on `this` also has some other weird consequences. Take this program:

```dart
class Person {
  name { "Fred" }
  greet {
    IO.print("Hi, my name is", name)
  }
}

(new Person).greet
```

Right now, it works just fine. But if you change it to:

```dart
var name = "WTF" // <-- Add this.

class Person {
  name { "Fred" }
  greet {
    IO.print("Hi, my name is", name)
  }
}

(new Person).greet
```

Now, it prints `Hi, my name is WTF`, which I don't think is what users intuit. (Though, for what it's worth, Dart *does* work this way.)

The obvious solution is that when you see an identifier in a class, you only look for locals in the method. If it isn't defined there, you can consider it a getter on `this` and don't look outside of the class. Alas, that doesn't work:

```dart
class Foo {
  method {
    IO.print("The String class is", String)
  }
}

(new Foo).method
```

Here, we'd compile `String` as `this.String`, which is clearly not right!

Distilled down, the two problems are:

1. How do we handle top-level variables defined after their use?
2. How do we decide if an identifier in a method is a variable outside of the class or a getter on `this`?

I'll leave this bug for the problem, and we can use separate issues for proposals to address these. (I'm filing both of these problems as a single bug because they're intertwined such that any solution will have to solve both.)

**Comment by zeckalpha** [comment]
I'm not sure this can be resolved while keeping single-pass compilation.

**Comment by munificent** [comment]
I have two ideas that can address this (while still being single-pass). I'll try to get some time to write them up tonight.

**Comment by munificent** [comment]
OK!

I got the top level names proposal we talked about (#106) implemented. It seems to work just fine. When I first changed the scoping rules, only three tests failed, which gives me some confidence that this change is intuitive and doesn't break any behavior that was being relied on.

Added lots of new tests for the new behavior.

One corner I'm still iffy about is what happens if you use a forward declared variable before it's defined, like:

```dart
IO.print(Foo)
var Foo = "defined"
```

I'd initially decided that should be a runtime error. I can do that, but it does have a runtime cost and adds some complexity to the implementation.

For now, implicitly declared variables are initialized to `null`. This is consistent with how uninitialized fields work, so I like that.
